File: travel-advisor/src/main/java/com/shadoworacle/travel_advisor/services/RouteOptimizationService.java
package com.shadoworacle.travel_advisor.services;

import com.shadoworacle.travel_advisor.models.RouteOptimization;
import com.shadoworacle.travel_advisor.models.Trip;
import com.shadoworacle.travel_advisor.repositories.RouteOptimizationRepository;
import com.shadoworacle.travel_advisor.repositories.TripRepository;
import jakarta.persistence.EntityNotFoundException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class RouteOptimizationService {

    private final RouteOptimizationRepository routeOptimizationRepository;
    private final TripRepository tripRepository;
        
    // TODO: Add any additional dependencies for ML integration

    @Autowired
    public RouteOptimizationService(RouteOptimizationRepository routeOptimizationRepository, 
                                    TripRepository tripRepository) {
        this.routeOptimizationRepository = routeOptimizationRepository;
        this.tripRepository = tripRepository;
    }

    public List<RouteOptimization> getAllRouteOptimizations() {
        return routeOptimizationRepository.findAll();
    }

    public RouteOptimization getRouteOptimizationById(Long id) {
        return routeOptimizationRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Route optimization not found with id: " + id));
    }

    public Optional<RouteOptimization> getRouteOptimizationByTripId(Long tripId) {
        return routeOptimizationRepository.findByTripId(tripId);
    }

    public List<RouteOptimization> getRouteOptimizationsByType(String optimizationType) {
        return routeOptimizationRepository.findByOptimizationType(optimizationType);
    }

    public List<RouteOptimization> getRouteOptimizationsByCriteria(String criteria) {
        return routeOptimizationRepository.findByOptimizationCriteria(criteria);
    }

    public RouteOptimization createRouteOptimization(RouteOptimization routeOptimization) {
        // Verify trip exists
        Long tripId = routeOptimization.getTrip().getId();
        Trip trip = tripRepository.findById(tripId)
                .orElseThrow(() -> new EntityNotFoundException("Trip not found with id: " + tripId));
        
        // Check if a route optimization already exists for this trip
        Optional<RouteOptimization> existingOptimization = routeOptimizationRepository.findByTripId(tripId);
        if (existingOptimization.isPresent()) {
            throw new IllegalStateException("A route optimization already exists for trip with id: " + tripId);
        }

        // Set the trip
        routeOptimization.setTrip(trip);
        
        // Call ML model to generate the optimized route
        optimizeRoute(routeOptimization);
        
        return routeOptimizationRepository.save(routeOptimization);
    }

    public RouteOptimization updateRouteOptimization(Long id, RouteOptimization routeOptimization) {
        // Check if route optimization exists
        RouteOptimization existingOptimization = getRouteOptimizationById(id);
        
        // Set ID to ensure we're updating the existing record
        routeOptimization.setId(id);
        
        // Preserve the trip relationship
        routeOptimization.setTrip(existingOptimization.getTrip());
        
        // // Re-run optimization if criteria changed
        // if (!existingOptimization.getOptimizationCriteria().equals(routeOptimization.getOptimizationCriteria()) ||
        //     !existingOptimization.getOptimizationType().equals(routeOptimization.getOptimizationType())) {
        //     optimizeRoute(routeOptimization);
        // }
        
        return routeOptimizationRepository.save(routeOptimization);
    }

    public void deleteRouteOptimization(Long id) {
        // Check if route optimization exists
        getRouteOptimizationById(id);
        routeOptimizationRepository.deleteById(id);
    }

    /**
     * Optimizes a route based on given criteria and constraints.
     * This method integrates with a Python-based ML model.
     */
    private void optimizeRoute(RouteOptimization routeOptimization) {
        // ===== MACHINE LEARNING INTEGRATION POINT =====
        
        // TODO: Implement Python ML model integration using one of the following approaches:
        
        // OPTION 1: Use ProcessBuilder to call a Python script
        /*
        try {
            // Prepare data for the Python script
            String tripData = convertTripToJson(routeOptimization.getTrip());
            String optimizationCriteria = routeOptimization.getOptimizationCriteria();
            
            // Create temporary file with trip data if needed
            // ...
            
            // Execute Python script
            ProcessBuilder processBuilder = new ProcessBuilder("python", 
                "path/to/route_optimizer.py", 
                "--trip-data", tripData,
                "--criteria", optimizationCriteria);
                
            Process process = processBuilder.start();
            
            // Read results from the Python script
            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream()))) {
                String line;
                StringBuilder output = new StringBuilder();
                while ((line = reader.readLine()) != null) {
                    output.append(line).append("\n");
                }
                
                // Parse the output and update the RouteOptimization object
                parseAndUpdateOptimizationResults(routeOptimization, output.toString());
            }
            
            int exitCode = process.waitFor();
            if (exitCode != 0) {
                throw new RuntimeException("Route optimization failed with exit code: " + exitCode);
            }
        } catch (Exception e) {
            throw new RuntimeException("Error running route optimization", e);
        }
        */
        
        // OPTION 2: Use REST API to a separate Python service
        /*
        try {
            // Prepare request data
            String requestBody = convertTripToJson(routeOptimization.getTrip());
            
            // Make HTTP request to Python service
            HttpClient client = HttpClient.newHttpClient();
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create("http://localhost:5000/optimize-route"))
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(requestBody))
                    .build();
                    
            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
            
            if (response.statusCode() != 200) {
                throw new RuntimeException("Route optimization service returned error code: " + response.statusCode());
            }
            
            // Parse the response and update the RouteOptimization object
            parseAndUpdateOptimizationResults(routeOptimization, response.body());
        } catch (Exception e) {
            throw new RuntimeException("Error calling route optimization service", e);
        }
        */
        
        // For now, use placeholder values
        routeOptimization.setTotalDistance(100.0);
        routeOptimization.setTotalTravelTimeMinutes(120);
        
        // Placeholder route data
        routeOptimization.setRouteJson("{}"); // Store route data as JSON string
    }
    
    /**
     * Parses the output from the ML model and updates the RouteOptimization object.
     */
    private void parseAndUpdateOptimizationResults(RouteOptimization routeOptimization, String mlOutput) {
        // TODO: Implement parsing logic based on your ML model's output format
        
        // Example implementation:
        /*
        try {
            JSONObject json = new JSONObject(mlOutput);
            
            double distance = json.getDouble("total_distance_km");
            int travelTime = json.getInt("total_travel_time_minutes");
            String routeData = json.getJSONArray("route_details").toString();
            
            routeOptimization.setTotalDistanceInKm(distance);
            routeOptimization.setTotalTravelTimeInMinutes(travelTime);
            routeOptimization.setRouteData(routeData);
        } catch (JSONException e) {
            throw new RuntimeException("Error parsing ML model output", e);
        }
        */
    }
}

File: travel-advisor-frontend/src/services/routeOptimizationService.js
// src/services/routeOptimizationService.js
import api from './api';

// Create a service object with methods for each route optimization API operation
const routeOptimizationService = {
  /**
   * Get all route optimizations
   * @returns {Promise} Promise that resolves to an array of route optimizations
   */
  getAllRouteOptimizations: async () => {
    return api.get('/route-optimizations');
  },
  
  /**
   * Get route optimization by ID
   * @param {number} id - The route optimization ID
   * @returns {Promise} Promise that resolves to a route optimization object
   */
  getRouteOptimizationById: async (id) => {
    return api.get(`/route-optimizations/${id}`);
  },
  
  /**
   * Get route optimization for a specific trip
   * @param {number} tripId - The trip ID
   * @returns {Promise} Promise that resolves to a route optimization object
   */
  getRouteOptimizationByTripId: async (tripId) => {
    return api.get(`/route-optimizations/trip/${tripId}`);
  },
  
  /**
   * Get route optimizations by type
   * @param {string} type - The optimization type
   * @returns {Promise} Promise that resolves to an array of route optimizations
   */
  getRouteOptimizationsByType: async (type) => {
    return api.get(`/route-optimizations/type/${type}`);
  },
  
  /**
   * Get route optimizations by criteria
   * @param {string} criteria - The optimization criteria (e.g., DISTANCE, TIME, SCENIC)
   * @returns {Promise} Promise that resolves to an array of route optimizations
   */
  getRouteOptimizationsByCriteria: async (criteria) => {
    return api.get(`/route-optimizations/criteria/${criteria}`);
  },
  
  /**
   * Create a new route optimization
   * @param {Object} routeOptimizationData - The route optimization data
   * @returns {Promise} Promise that resolves to the created route optimization
   */
  createRouteOptimization: async (routeOptimizationData) => {
    return api.post('/route-optimizations', routeOptimizationData);
  },
  
  /**
   * Update an existing route optimization
   * @param {number} id - The route optimization ID to update
   * @param {Object} routeOptimizationData - The updated route optimization data
   * @returns {Promise} Promise that resolves to the updated route optimization
   */
  updateRouteOptimization: async (id, routeOptimizationData) => {
    return api.put(`/route-optimizations/${id}`, routeOptimizationData);
  },
  
  /**
   * Delete a route optimization
   * @param {number} id - The route optimization ID to delete
   * @returns {Promise} Promise that resolves when the route optimization is deleted
   */
  deleteRouteOptimization: async (id) => {
    return api.delete(`/route-optimizations/${id}`);
  },
  
  /**
   * Reoptimize a route with new parameters
   * @param {number} id - The route optimization ID
   * @param {string} [criteria] - New optimization criteria
   * @param {string} [type] - New optimization type
   * @returns {Promise} Promise that resolves to the reoptimized route
   */
  reoptimizeRoute: async (id, criteria, type) => {
    return api.post(`/route-optimizations/${id}/reoptimize`, null, {
      params: { criteria, type }
    });
  },
  
  /**
   * Compare different optimization strategies for a trip
   * @param {number} tripId - The trip ID
   * @param {string} strategies - Comma-separated list of strategies to compare
   * @returns {Promise} Promise that resolves to the comparison results
   */
  compareOptimizationStrategies: async (tripId, strategies = 'DISTANCE,TIME,SCENIC') => {
    return api.get('/route-optimizations/compare', {
      params: { tripId, strategies }
    });
  }
};

export default routeOptimizationService;